### PR TITLE
[ADD] Module <product_variant_auto_creation_selectable> from odoomrp-wip

### PR DESCRIPTION
**Pull request by oihane**

### CODE REVIEW

**Comment by oihane** [comment]
Travis fails with message:

```
File "/home/travis/build/OCA/product-variant/product_variants_no_automatic_creation/models/pricelist.py", line 6, in <module>
from openerp import models, fields, tools, api, exceptions, _
ImportError: cannot import name models
----------
Couldn't load module product_variants_no_automatic_creation
cannot import name models
Modules loaded message not found.
```

**Comment by pedrobaeza** [comment]
Make a rebase

**Comment by oihane** [comment]
Done!

File: product_variants_no_automatic_creation/README.rst [hunk 1]
```diff
@@ -0,0 +1,46 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :alt: License: AGPL-3
+
+Product Variants
+================
+
+This module creates a check in categories and an option in product templates,
+so that it does not create the product variants when the attributes are
+assigned. This is used by those modules that create the product when it is
+estrictly necessary.
```
Comment (pedrobaeza) [inline product_variants_no_automatic_creation/README.rst]: s/estrictly necessary/strictly needed

File: product_variants_no_automatic_creation/README.rst [hunk 2]
```diff
@@ -0,0 +1,46 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :alt: License: AGPL-3
+
+Product Variants
```
Comment (pedrobaeza) [inline product_variants_no_automatic_creation/README.rst]: `Prevents the automatic creation of the variants of a product`

File: product_variants_no_automatic_creation/README.rst (hunk 1 above)
Comment (pedrobaeza) [inline product_variants_no_automatic_creation/README.rst]: Explain in the "Usage" section which options are available.

File: product_variants_no_automatic_creation/__openerp__.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see http://www.gnu.org/licenses/.
+#
+##############################################################################
+
+{
+    "name": "Product Variants",
+    "summary": "Disable automatic product variant creation",
+    "version": "1.0",
+    "category": "Hidden/Dependency",
```
Comment (pedrobaeza) [inline product_variants_no_automatic_creation/__openerp__.py]: Product Management

File: product_variants_no_automatic_creation/README.rst (hunk 2 above)
Comment (oihane) [inline product_variants_no_automatic_creation/README.rst]: Is it a name suggestion?

File: product_variants_no_automatic_creation/__openerp__.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see http://www.gnu.org/licenses/.
+#
+##############################################################################
+
+{
+    "name": "Product Variants",
```
Comment (pedrobaeza) [inline product_variants_no_automatic_creation/__openerp__.py]: This can be the same text as in the title of README

File: product_variants_no_automatic_creation/README.rst (hunk 2 above)
Comment (pedrobaeza) [inline product_variants_no_automatic_creation/README.rst]: Well, more or less

File: product_variants_no_automatic_creation/models/pricelist.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+##############################################################################
+
+from openerp import models, fields, tools, api, exceptions, _
+
+
+class ProductPricelist(models.Model):
+    _inherit = 'product.pricelist'
+
+    @api.model
+    def _price_rule_get_multi(self, pricelist, products_by_qty_by_partner):
```
Comment (pedrobaeza) [inline product_variants_no_automatic_creation/models/pricelist.py]: Why is this here?

**Comment by pedrobaeza** [comment]
Do you know what to test?

File: product_variants_no_automatic_creation/models/pricelist.py (hunk 5 above)
Comment (oihane) [inline product_variants_no_automatic_creation/models/pricelist.py]: :sweat: I do not remember why... we've done modules for sale, purchase and mrp, I think they need this one, but I can't remember why...

File: product_variants_no_automatic_creation/README.rst [hunk 6]
```diff
@@ -0,0 +1,46 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :alt: License: AGPL-3
+
+Prevent the automatic creation of the variants of a product
```
Comment (pedrobaeza) [inline product_variants_no_automatic_creation/README.rst]: s/Prevent/Prevents

File: product_variants_no_automatic_creation/models/pricelist.py (hunk 5 above)
Comment (pedrobaeza) [inline product_variants_no_automatic_creation/models/pricelist.py]: We should investigate it then

File: product_variants_no_automatic_creation/README.rst (hunk 6 above)
Comment (oihane) [inline product_variants_no_automatic_creation/README.rst]: Are you sure, it is the title...

**Comment by pedrobaeza** [comment]
Another thing that I notice now: the name must be in singular to follow this convention. Please rename it.

File: product_variants_no_automatic_creation/README.rst (hunk 6 above)
Comment (pedrobaeza) [inline product_variants_no_automatic_creation/README.rst]: (The module) prevents the automatic...

Comment (pedrobaeza) [inline product_variants_no_automatic_creation/README.rst]: The infinitive can be used in case of an imperative. The title can also be interpreted that way maybe...

**Comment by pedrobaeza** [comment]
Talking directly with @oihane, we agree to call it _product_variant_auto_creation_selectable_

**Comment by oihane** [comment]
@pedrobaeza take a look again please, I think it is ready to delete the **Work in progress** label

File: product_variant_auto_creation_selectable/README.rst [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+Product variant automatic creation selectable
+=============================================
+
+This module creates a check in categories and an option in product templates,
+so that it does not create the product variants when the attributes are
+assigned. This is used by those modules that create the product when it is
+strictly needed.
+
+
```
Comment (pedrobaeza) [inline product_variant_auto_creation_selectable/README.rst]: It still lacks the Usage section explaining possible values for product categories and templates.

File: product_variant_auto_creation_selectable/models/pricelist.py [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+
+from openerp import models, fields, tools, api, exceptions, _
+
+
+class ProductPricelist(models.Model):
+    _inherit = 'product.pricelist'
+
+    @api.model
+    def _price_rule_get_multi(self, pricelist, products_by_qty_by_partner):
+        context = self.env.context.copy() or {}
```
Comment (pedrobaeza) [inline product_variant_auto_creation_selectable/models/pricelist.py:13]: We still don't know for what is this needed

Comment (pedrobaeza) [inline product_variant_auto_creation_selectable/models/pricelist.py:13]: Also this is a hard-copy. We should try to avoid it.

**Comment by pedrobaeza** [comment]
One last thing: guys from Xpansa tell me that they have a wizard to create manually product variants, so it's worth to wait to integrate this wizard in this module. I let the WIP label until then.

**Comment by pedrobaeza** [comment]
@oihane, please include this commit that ensures the uniqueness of product/attribute combination: https://github.com/odoomrp/odoomrp-wip/commit/e618fcd6b524b27f93d973168c437d8234060550

**Comment by oihane** [comment]
@pedrobaeza included the commit you commented, and also added some changes suggested in https://github.com/OCA/purchase-workflow/pull/146 (try me button and version)

**Comment by pedrobaeza** [comment]
Thanks, we are waiting for response from Xpansa. If not, we will develop this feature.

File: product_variant_auto_creation_selectable/models/pricelist.py (hunk 8 above)
Comment (oihane) [inline product_variant_auto_creation_selectable/models/pricelist.py:13]: I think I found why is this required.

As an example, when working in sale with templates (no variant created) it does not know how much does it cost so it gets the template cost and the price extra from the attributes to provide a price unit.

Comment (pedrobaeza) [inline product_variant_auto_creation_selectable/models/pricelist.py:13]: Uhm, I see, but I don't like this code at all. We have to look for another possibility, or provide hooks in Odoo core in a PR.

**Comment by pedrobaeza** [comment]
@oihane, what is the status of this? Is all done and can I review it? Can you please squash commits so that each author has only a few commits about relevant things (avoiding other fixes/patches)?

**Comment by oihane** [comment]
There are some things missing, but you can review (one of them is README title). The other thing is related to the price, that you said you don't like. And maybe you shoul take a look in runbot, to see how the "configurator" is shown.

**Comment by oihane** [comment]
Do I have to add a translated es.po?

**Comment by pedrobaeza** [comment]
It would be good to have spanish translations from the beginning. I'll check the rest. Why don't you make the PR as GitHub odoomrp team and we can both make commits to that branch?

File: product_variant_auto_creation_selectable/models/product.py [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+        if (not values.get('attribute_value_ids') and
+                values.get('product_attributes')):
+            values['attribute_value_ids'] = (
+                (4, x[2]['value'])
+                for x in values.get('product_attributes'))
+        return super(ProductProduct, self).create(values)
+
+    @api.one
+    def write(self, values):
+        return super(ProductProduct, self).write(values)
```
Comment (pedrobaeza) [inline product_variant_auto_creation_selectable/models/product.py]: @oihane, for what is this needed?

File: product_variant_auto_creation_selectable/models/product.py [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+
+
+class ProductProduct(models.Model):
+    _inherit = 'product.product'
+
+    product_attributes = fields.One2many(
+        comodel_name='product.product.attribute', inverse_name='product',
+        string='Product attributes', copy=False)
+
+    @api.one
```
Comment (pedrobaeza) [inline product_variant_auto_creation_selectable/models/product.py]: Remove this line. In onchange is useless

File: product_variant_auto_creation_selectable/models/product.py [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+         ('empty', 'Use the category value')],
+        string='Variant creation', required=True, default='empty',
+        help="This selection defines if variants for all attribute "
+             "combinations are going to be created automatically at saving "
+             "time.")
+
+    @api.multi
+    @api.onchange('no_create_variants')
+    def onchange_no_create_variants(self):
+        self.ensure_one()
```
Comment (pedrobaeza) [inline product_variant_auto_creation_selectable/models/product.py]: This one is useless here.

File: product_variant_auto_creation_selectable/models/product.py [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+    no_create_variants = fields.Selection(
+        [('yes', "Don't create them automatically"),
+         ('no', "Create them automatically"),
+         ('empty', 'Use the category value')],
+        string='Variant creation', required=True, default='empty',
+        help="This selection defines if variants for all attribute "
+             "combinations are going to be created automatically at saving "
+             "time.")
+
+    @api.multi
```
Comment (pedrobaeza) [inline product_variant_auto_creation_selectable/models/product.py]: This one is useless here.

**Comment by pedrobaeza** [comment]
Note to myself: for avoiding compatibility problems, default value in test mode should be "create variants", and tests should explicitly indicate the choice to test. This way, tests that expects to have the variants created don't fail.

File: product_variant_auto_creation_selectable/models/product_attribute_configurator.py [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+    price_extra = fields.Float(
+        compute='_get_price_extra', string='Attribute Price Extra',
+        digits=dp.get_precision('Product Price'),
+        help="Price Extra: Extra price for the variant with this attribute"
+        " value on sale price. eg. 200 price extra, 1000 + 200 = 1200.")
+
+
+class ProductProductAttribute(models.Model):
+    _inherit = 'product.attribute.configurator'
+    _name = 'product.product.attribute'
```
Comment (pedrobaeza) [inline product_variant_auto_creation_selectable/models/product_attribute_configurator.py]: There is no security access file for this model.

File: product_variant_auto_creation_selectable/models/product.py [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+                        'value_id': value.id,
+                    })
+        result = self._get_act_window_dict(
+            'product_variants_auto_creation_selectable.attribute_price_action')
+        return result
+
+    @api.model
+    def name_search(self, name='', args=None, operator='ilike', limit=100):
+        # Make a search with default criteria
+        temp = super(models.Model, self).name_search(
```
Comment (pedrobaeza) [inline product_variant_auto_creation_selectable/models/product.py]: Redo this using expresion.OR
